packages/flutter/lib/src/widgets/interactive_viewer.dart

```diff
@@ -0,0 +1,1136 @@
+// and rotation in the gesture, if any. Scale starts at 1 and rotation
+// starts at 0. Translate will have 0 scale and 0 rotation because it uses
+// only one finger.
+_GestureType _getGestureType(double scale, double rotation) {
+  if ((scale - 1).abs() > rotation.abs()) {
+    return _GestureType.scale;
+  } else if (rotation != 0) {
+    return _GestureType.rotate;
+  } else {
+    return _GestureType.translate;
+  }
+}
+
+// Return the translation from the given Matrix4 as an Offset.
+Offset _getMatrixTranslation(Matrix4 matrix) {
+  final Vector3 nextTranslation = matrix.getTranslation();
+  return Offset(nextTranslation.x, nextTranslation.y);
+}
+
+// Transform the four corners of the viewport by the inverse of the given
+// matrix. This gives the viewport after the child has been transformed by the
+// given matrix. The viewport transforms as the inverse of the child (i.e.
+// moving the child left is equivalent to moving the viewport right).
+Quad _transformViewport(Matrix4 matrix, Rect viewport) {
+  final Matrix4 inverseMatrix = matrix.clone()..invert();
+  return Quad.points(
+    inverseMatrix.transform3(Vector3(
+      viewport.topLeft.dx,
+      viewport.topLeft.dy,
+      0.0,
+    )),
+    inverseMatrix.transform3(Vector3(
+      viewport.topRight.dx,
+      viewport.topRight.dy,
+      0.0,
+    )),
+    inverseMatrix.transform3(Vector3(
+      viewport.bottomRight.dx,
+      viewport.bottomRight.dy,
+      0.0,
+    )),
+    inverseMatrix.transform3(Vector3(
+      viewport.bottomLeft.dx,
+      viewport.bottomLeft.dy,
+      0.0,
+    )),
+  );
+}
+
+// Get the size of the child given its RenderBox and the viewport's Size.
+//
+// In some cases (i.e. a Table that's wider and/or taller than the device),
+// renderBox.size will give the size of the device, even though the child is
+// drawn beyond the viewport. The intrinsic size can then be used to set the
+// boundary to the full size of the child.
+//
+// In other cases (i.e. an Image whose original size is larger than the
+// viewport but is being fit to the viewport), renderBox.size will also give
+// the size of the viewport, and the boundary should remain at the viewport. In
+// that case, the intrinsic size is not used.
+Size _getChildSize(RenderBox renderBox, BoxConstraints constraints) {
+  double width = renderBox.size.width;
+  double height = renderBox.size.height;
+  final double minIntrinsicWidth = renderBox.getMinIntrinsicWidth(constraints.maxHeight);
+  final double maxIntrinsicWidth = renderBox.getMaxIntrinsicWidth(constraints.maxHeight);
+  final double minIntrinsicHeight = renderBox.getMinIntrinsicHeight(constraints.maxWidth);
+  final double maxIntrinsicHeight = renderBox.getMaxIntrinsicHeight(constraints.maxWidth);
+
+  if (minIntrinsicWidth == maxIntrinsicWidth) {
+    width = minIntrinsicWidth;
+  }
+  if (minIntrinsicHeight == maxIntrinsicHeight) {
+    height = minIntrinsicHeight;
+  }
```

Comment:
I don't fully understand the intend of this. Why do we use a different child width/height if min and max intrinsic width/height are the same?

Reply:
I did some research about the whole problem with _getChildSize from several comments here.  This is how I believe users will expect InteractiveViewer to work:

  1. If a user passes a child that has an explicit size larger than the InteractiveViewer itself, it should be drawn at its explicit size and should be able to have overflowing parts panned into view. For example: `Container(width: 10000, height: 10000)`
  1. If a user passes a child without an explicit size, it should size itself as it would if the InteractiveViewer weren't there.  For example, a Container with no width/height should fill the InteractiveViewer. An image should be able to use BoxFit like normal.

For point number 1, I can't use `renderBox.size` to get the true size of the child when it's larger than the screen.  The value it gives will be clamped to the size of the screen.  Wrapping it in an OverflowBox doesn't help, as `renderBox.size` still gives the size of the screen.  Wrapping it in two SingleChildScrollViews does work and renderBox.size gives the correct size, if I use one with axis horizontal and one with vertical, and use NeverScrollableScrollPhysics to prevent actual scrolling.  Besides being hacky though, this breaks point number 2, because the child no longer receives any size constraints.

My current implementation of _getChildSize does work for the cases I've given, but is there a more robust way to achieve this?

Reply:
The getFooIntrinsicBar() methods are quite expensive because they potentially have to compute the layout of the entire subtree.  Would constraints.hasTightWidth and constraints.hasTightHeight work as well here? It's not quite the same, but it might have the same effect for the use-cases that matter.